### PR TITLE
Test/rate limiters

### LINKED ISSUE #528: [TEST] Jax Rate Limiters
### What do you want to test?
Jax rate limiter components

### Outline of test structure
* Unit tests
* Test components and hooks

### Definition of done
Passing checks, cover all hooks, edge cases considered

### Mandatory checklist before making a PR
* [ ] The success criteria laid down in “Definition of done” are met.
* [ ] Test code is documented - docstrings for methods and classes, static types for arguments.
* [ ] Documentation is updated - README, CONTRIBUTING, or other documentation.

### PR DESCRIPTION
**Pull request by mmorris44**
## What?
Test Jax `RateLimiter` components.
## Why?
Need comprehensive tests for `jax` system components.
## How?
- Tested `MinSizeRateLimiter` and `SampleToInsertRateLimiter` by manually calling hooks.
- Skipped tests which require higher versions of `dm-reverb`
## Extra
Closes #528.

### CODE REVIEW

**Review by RuanJohn — APPROVED** [review]
Looks good. Thank you @mmorris44. We defintely need to revisit this test when `reverb` is upgraded.

**Review by RuanJohn — CHANGES_REQUESTED** [review]
@mmorris44 please see [here](https://github.com/instadeepai/Mava/blob/2b81ee48a1c175536452e6ee7aeb2f6a2088d2fe/setup.py#L29), the reverb version has been upgraded.

**Comment by mmorris44** [comment]
> @mmorris44 please see [here](https://github.com/instadeepai/Mava/blob/2b81ee48a1c175536452e6ee7aeb2f6a2088d2fe/setup.py#L29), the reverb version has been upgraded.

Where's the oops emoji when you need it?

**Comment by RuanJohn** [comment]
The issue appears to be a reverb version. When I have `dm-reverb==0.7.2` installed the tests are able run with everything uncommented, but not for `dm-reverb==0.7.3`.

**Comment by mmorris44** [comment]
@RuanJohn any updates on whether the tests work in Docker for you? Still failing on my end

**Review by RuanJohn — DISMISSED** [review]
Thank you @mmorris44 :+1:

File: tests/jax/components/building/rate_limiters_test.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    sample_to_insert_rate_limiter.on_building_data_server_rate_limiter(builder)
+
+    reverb_rate_limiter = builder.store.rate_limiter_fn()
+    assert isinstance(reverb_rate_limiter, reverb.rate_limiters.SampleToInsertRatio)
+
+    # TODO: remove below when following test is no longer skipped
+    assert repr(reverb_rate_limiter).split("min_size_to_sample=")[1][:3] == "100"
+    assert repr(reverb_rate_limiter).split("samples_per_insert=")[1][:2] == "16"
+    assert repr(reverb_rate_limiter).split("min_diff_=")[1][:4] == "1100"
+    assert repr(reverb_rate_limiter).split("max_diff=")[1][:4] == "2100"
```
Comment (OmaymaMahjoub) [inline tests/jax/components/building/rate_limiters_test.py:93]: Can we make it more  explicit of how min_diff and max_diff 1100 and 2100 by making the equation str like
`offset=sample_to_insert_rate_limiter.config.min_data_server_size*sample_to_insert_rate_limiter.config.samples_per_insert`
`repr(reverb_rate_limiter).split("min_diff_=")[1][:4] == str(int(offset-sample_to_insert_rate_limiter.config.error_buffer))`

File: tests/jax/components/building/rate_limiters_test.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    sample_to_insert_rate_limiter.on_building_data_server_rate_limiter(builder)
+
+    reverb_rate_limiter = builder.store.rate_limiter_fn()
+    assert isinstance(reverb_rate_limiter, reverb.rate_limiters.SampleToInsertRatio)
+
+    # TODO: remove below when following test is no longer skipped
+    assert repr(reverb_rate_limiter).split("min_size_to_sample=")[1][:3] == "100"
+    assert repr(reverb_rate_limiter).split("samples_per_insert=")[1][:2] == "16"
+    assert repr(reverb_rate_limiter).split("min_diff_=")[1][:4] == "1440"
+    assert repr(reverb_rate_limiter).split("max_diff=")[1][:4] == "1760"
```
Comment (OmaymaMahjoub) [inline tests/jax/components/building/rate_limiters_test.py:137]: The same comment: can we make it more explicit to be more understandable.

**Review by OmaymaMahjoub — CHANGES_REQUESTED** [review]
Looks good thanks for the PR!
Just the remove of the following methods
`test_min_size_rate_limiter_attributes`
`test_sample_to_insert_rate_limiter_with_error_buffer_attributes`
`test_sample_to_insert_rate_limiter_no_error_buffer_attributes`

**Comment by mmorris44** [comment]
> Looks good thanks for the PR! Just we need to remove of the following methods `test_min_size_rate_limiter_attributes` `test_sample_to_insert_rate_limiter_with_error_buffer_attributes` `test_sample_to_insert_rate_limiter_no_error_buffer_attributes`

I've kept the tests in there for now (but skipped) so that we can switch back to them when reverb is upgraded. The way we are currently doing it is very messy.

File: tests/jax/components/building/rate_limiters_test.py (hunk 1 above)
Comment (mmorris44) [reply tests/jax/components/building/rate_limiters_test.py:93]: Done!

File: tests/jax/components/building/rate_limiters_test.py (hunk 2 above)
Comment (mmorris44) [reply tests/jax/components/building/rate_limiters_test.py:137]: Also done

**Review by RuanJohn — COMMENTED** [review]
Did you mean for the component dependency map to be included?

**Review by OmaymaMahjoub — DISMISSED** [review]
Looks good to me, thanks Matthew!

**Comment by mmorris44** [comment]
> Did you mean for the component dependency map to be included?

Accidentally added, good catch! Thanks @RuanJohn. Have removed it now.

**Review by RuanJohn — APPROVED** [review]

**Review by OmaymaMahjoub — DISMISSED** [review]
All good from my side!

**Review by OmaymaMahjoub — APPROVED** [review]

**Review by RuanJohn — APPROVED** [review]
